Add format checking for integer and number types.

CheesyBoy123 opened the pull request:
Motivation:

In order to resolve https://github.com/eclipse-vertx/vertx-json-schema/issues/124 I've added format checking for integer and number types. These types will now allow the following formats:
type = integer, format = int32
type = integer, format = int64
type = number, format = float
type = number, format = double

These formats are optional, so if the user doesn't specify the format OR the format doesn't match the above then it is ignored.

After installing the snapshot of this locally, it also resolves the reproducer for the related issue - https://github.com/eclipse-vertx/vertx-openapi/pull/51

Conformance:

You should have signed the Eclipse Contributor Agreement as explained in https://github.com/eclipse/vert.x/blob/master/CONTRIBUTING.md ✅ 
Please also make sure you adhere to the code style guidelines: https://github.com/vert-x3/wiki/wiki/Vert.x-code-style-guidelines ✅

pk-work commented:
Hi @CheesyBoy123 first of all thanks for your contribution! Sorry, that the feedback took so long.

I have a question. According to the OpenAPI spec, an integer of format int32 could by any value between −2,147,483,648 and 2,147,483,647 [1], would you agree?

In my opinion the value "32767" is a valid int32 representation, which is MAX_VALUE of Short. But according to your implementation any other Type than `Integer` is invalid. I think the current check is not sufficient.

I think the same is valid for format `float` and `double`, a float object could represent a valid double value and vice versa.

I'm also not sure if we should define these format checks in `vertx-json-schema`, because as far as I know the JSON Schema spec does not define formats for integers or numbers. The formats we talk about come from OAS [2], so we should check them in vertx-openapi.

I know you started this fix in vertx-openapi and then moved into vertx-json-schema, but I think vertx-openapi is the better way to check this. Because in vertx-openapi we transform the value representation which is always a string into a Java type. If the format is e.g. int32, we can check if it is possible to transform the String into an Integer. If not because the number represented by the String is bigger than Integer.MAX_VALUE we can throw an error.

What do you think?

[1] https://spec.openapis.org/registry/format/int32.html
[2] https://spec.openapis.org/registry/format/index.html#values

pk-work commented:
Hi @CheesyBoy123 ,

these are steps from a request parameter to the validation. I just wanted to show them that we are all on the same page.

1. (Vert.x OpenAPI) ParameterTransformer transforms the rawValue (String) of a parameter with Jackson into an object [1]
2. (Vert.x OpenAPI) RequestValidator passes the transformedValue into (Vert.x Json Schema) Validator [2]
3. (Vert.x Json Schema) Validator checks the format.

As you can see, Jackson decides if a String gets converted into an Integer or Long. Jackson doesn't consider if the format is `int32` or `int64`. Because of this, your check [3] would reject all values <= Integer.Max if they have the format `int64`. So effectively you only check if a number is in a given range. These kind of check you can also perform with Range [4].

I just checked it, but Jackson always converts a floating-point number into a Double. So your check for format "float" [5] could never be true, in case you use Vert.x OpenAPI.

Maybe I oversee something, but I think we need a different kind of check, or interfere the transformation strep.

I would suggest that format "int64" also accepts all integers below Integer.Max.
I don't yet have a good answer for the precision problem, but maybe we can simply accept it.

But nevertheless we can think about an API which allows you to plugin a custom format validator. Since `format` is not really described, this is a good idea.

What do you think?

[1] https://github.com/eclipse-vertx/vertx-openapi/blob/ff26687aa607adbf25c65e14f9af70cfc0f8ab4d/src/main/java/io/vertx/openapi/validation/transformer/ParameterTransformer.java#L37
[2] https://github.com/eclipse-vertx/vertx-openapi/blob/ff26687aa607adbf25c65e14f9af70cfc0f8ab4d/src/main/java/io/vertx/openapi/validation/impl/RequestValidatorImpl.java#L126
[3] https://github.com/eclipse-vertx/vertx-json-schema/pull/125/files#diff-781b21b8bc3101429fbb4523374f05282697d15f549a4d1a3a21ccec60529281R269
[4] https://json-schema.org/understanding-json-schema/reference/numeric#range
[5] https://github.com/eclipse-vertx/vertx-json-schema/pull/125/files#diff-781b21b8bc3101429fbb4523374f05282697d15f549a4d1a3a21ccec60529281R947

CheesyBoy123 commented:
Hey @pk-work let me convert this to being an interface that we can add into vertx-openapi. I think that will be the best bet forward for now.

CheesyBoy123 commented:
@pk-work I've made a quick update to create a new interface. In vertx-openapi  PR we are using that interface to validate the format of parameters/body. Let me know how this looks (I will add javadocs and such in a bit).

pk-work commented:
@CheesyBoy123 this looks good. With this approach everybody can implement custom format validations. We should also consider to add this new functionality to the docs [1].

[1] https://github.com/eclipse-vertx/vertx-json-schema/blob/master/src/main/asciidoc/index.adoc

CheesyBoy123 commented:
@pk-work i've updated the doc to reflect these changes! I've gone ahead and added java docs as well. Let me know if you think something else is missing! Thanks!

pk-work commented on src/main/java/io/vertx/json/schema/JsonFormatValidator.java:
```diff
@@ -0,0 +1,23 @@
+ */
+package io.vertx.json.schema;
+
+public abstract class JsonFormatValidator {
```
Is there a reason, why this isn't an interface?

CheesyBoy123 replied:
You're right this is better as an interface.

pk-work commented on src/main/java/examples/JsonSchemaExamples.java:
```diff
@@ -28,4 +29,14 @@ public void validate(JsonSchema schema, Object json) {
+    JsonSchemaOptions jsonSchemaOptionsWithFormatValidator = new JsonSchemaOptions()
+      .setBaseUri("https://vertx.io")
+      //Specify your own format validator here!
+      .setJsonFormatValidator(new DefaultJsonFormatValidatorImpl());
```
Maybe we should implement a dummy example here, could be implemented via a lambda.

pk-work commented on src/main/java/io/vertx/json/schema/JsonFormatValidator.java:
```diff
@@ -0,0 +1,23 @@
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+package io.vertx.json.schema;
+
```
I think this class must be annotated with `VertxGen`, as we are using it in `JsonSchemaOptions`, which is annotated with `DataObject`.

pk-work commented on src/main/java/io/vertx/json/schema/JsonSchemaOptions.java:
```diff
@@ -40,6 +41,11 @@ public class JsonSchemaOptions {
+  /**
+   * The json format validator that is used during schema validation.
+   */
+  private JsonFormatValidator jsonFormatValidator = new DefaultJsonFormatValidatorImpl();
```
Now when I think about this, it is might be a problem, because JsonSchemaOptions is annotated with JsonGen and we can't serialize the JsonFormatValidator.

And using `@GenIgnore` would make this unusable in Kotlin, etc .. If we cannot find a better solution, we could store the FQN and instantiate the class during deserialization. I have to double check with @vietj .

pk-work replied:
Okay, I talked to Julien. He made me aware of the builder pattern guidelines [1]. So we should not mix logic or algorithm with a DataObject. This means that we need an additional builder for SchemaRepository and Validator.

I will try to prepare a proposal this week.

[1] https://github.com/eclipse-vertx/vert.x/wiki/Builder-pattern-guidelines

pk-work commented on src/main/java/io/vertx/json/schema/JsonSchemaOptions.java:
```diff
@@ -40,6 +41,11 @@ public class JsonSchemaOptions {
+  /**
+   * The json format validator that is used during schema validation.
+   */
+  private JsonFormatValidator jsonFormatValidator = new DefaultJsonFormatValidatorImpl();
```
```suggestion
  private JsonFormatValidator jsonFormatValidator = (format, instance) -> null;
```

No need for `DefaultJsonFormatValidatorImpl`

pk-work commented on src/main/java/io/vertx/json/schema/impl/SchemaValidatorImpl.java:
```diff
@@ -955,6 +961,11 @@ private OutputUnit validate(final Object _instance, final JsonSchema schema, fin
       }
     }
 
+    String error = formatValidator.validateFormat((String)schema.get("format"), instance);
```
```suggestion
    String error = formatValidator.validateFormat((String) schema.get("format"), instance);
```

CheesyBoy123 commented on src/main/asciidoc/index.adoc:
```diff
@@ -104,10 +104,16 @@ To validate a schema:
 
-Adding custom formats is not allowed.This may sound like a limitation but it arises from the validation spec.
-Validators are expected to assume unknown formats as valid input, delegating to the application business code the role of validator for custom values.
+
```
Added a quick note that by default it is a NOOP and that unknown formats are treated as valid.

pk-work commented:
Merged with #128
